### PR TITLE
Resolve #1241, Resolve #1242 -- Early introduction of GameModes + Minor fixes

### LINKED ISSUE #1241: Gamemode Specific Map Scripting
Scripting of Gamemodes should be implemented.

### LINKED ISSUE #1242: Buffs Require Source
Because of this, buffs which normally are applied as the game starts (such as in URF) are unable to be applied.

### PR DESCRIPTION
**Pull request by cabeca1143**
* Map System Changes:
  * Map Scripting:
    * Organized into folders, each with a default `CLASSIC` gamemode.
    * Different Gamemodes are scripted in a separate class from `CLASSIC`, usually inheriting it. 
    * The map to which the Gamemode belongs is defined by it's Namespace.
        * **EX**: For Map1 = `MapScripts.Map1`, Map2 = `MapScripts.Map2` and so on for each map.
    * MapScripts:
      * Added `GlobalData` and `GameFeatures` to override certain aspects of the game (such as URF overriding Mana costs and CDR cap)
      * Introduced the URF Gamemode as demonstration
* General:
  * Config:
    * Changed the template `GameInfo` file to house a new `GameMode` term
  * Champions:
    * Level ups are done as the match starts, not during loading.
  * Buffs:
    * Changed packets to handle buffs without a `SourceSpell`.
  * Events:
    * Fixed OnKillUnit not being published when killing a Unit that doesn't reward gold.
  * Scripting:
    * Added `GetAllPlayers` and `GetAllChampions` functions.

### CODE REVIEW

File: GameServerConsole/Settings/GameInfo.json.template [hunk 1]
```diff
@@ -1,62 +1,63 @@
...
-        "dataPackage": "LeagueSandbox-Scripts"
-    },
-    "gameInfo": {
-        "MANACOSTS_ENABLED": false,
-        "COOLDOWNS_ENABLED": false,
-        "CHEATS_ENABLED": true,
-        "MINION_SPAWNS_ENABLED": false,
-        "CONTENT_PATH": "../../../../../Content",
-        "IS_DAMAGE_TEXT_GLOBAL": false
+  "players": [
```
Comment (lzardy) [inline GameServerConsole/Settings/GameInfo.json.template]: Maintain 4 space tab format.

File: GameServerConsole/Settings/GameInfo.json.template [hunk 2]
```diff
@@ -1,62 +1,63 @@
...
+        "27": 5289,
+        "28": 5335,
+        "29": 5335,
+        "30": 5335
+      }
     }
+  ],
+  "game": {
+    "map": 1,
+    "GameMode": "CLASSIC",
```
Comment (lzardy) [inline GameServerConsole/Settings/GameInfo.json.template]: Change to lowercase to match formatting.

File: GameServerLib/Config.cs [hunk 3]
```diff
@@ -23,13 +23,11 @@ public class Config
         public MapData MapData { get; private set; }
         public MapSpawns MapSpawns { get; private set; }
         public ContentManager ContentManager { get; private set; }
+        public IGameFeatures GameFeatures { get; set; }
```
Comment (lzardy) [inline GameServerLib/Config.cs]: Change to `private set` and utilize a function for setting.

File: GameServerConsole/Settings/GameInfo.json.template (hunk 1 above)
Comment (cabeca1143) [reply GameServerConsole/Settings/GameInfo.json.template]: Resolved

File: GameServerLib/Config.cs [hunk 4]
```diff
@@ -129,7 +129,12 @@ private string GetContentPath()
             return result;
         }
     }
-
+    public class GameFeatures : IGameFeatures
+    {
+        public bool CooldownsEnabled { get; set; }
```
Comment (lzardy) [inline GameServerLib/Config.cs]: Same as above, change to `private set` and utilize a function to set.
```suggestion
        public bool CooldownsEnabled { get; private set; }
        public bool ManaCostsEnabled { get; private set; }
        public bool MinionSpawnsEnabled { get; private set; }
```

File: GameServerLib/Content/CharData.cs [hunk 5]
```diff
@@ -4,27 +4,14 @@
...
 using GameServerCore.Domain;
 using GameServerCore.Enums;
+using GameServerLib.GameObjects.GlobalData;
 using LeagueSandbox.GameServer.Logging;
 using log4net;
 using Newtonsoft.Json;
 
 namespace LeagueSandbox.GameServer.Content
 {
     // TODO: Add Constants.var files to each Map's Content folder and assign values by reading them, currently this Data is only for Map1 as a placeholder.
```
Comment (lzardy) [inline GameServerLib/Content/CharData.cs]: Move comment to the new `GlobalData.cs` file.

File: GameServerLib/GameObjects/GlobalData.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using GameServerCore.Domain;
+
+namespace GameServerLib.GameObjects.GlobalData
```
Comment (lzardy) [inline GameServerLib/GameObjects/GlobalData.cs]: Move this file to `GameServerLib/Content` and change namespace.
```suggestion
namespace LeagueSandbox.GameServer.Content
```

File: GameServerLib/Maps/MapScriptHandler.cs [hunk 7]
```diff
@@ -557,7 +571,19 @@ public void AddFountain(TeamId team, Vector2 position)
         {
             FountainList.Add(team, new Fountain(_game, team, position, 1000));
         }
-
+        public List<IChampion> GetPlayers()
```
Comment (lzardy) [inline GameServerLib/Maps/MapScriptHandler.cs]: This function is unnecessary as scripts have access to ApiFunctionManager.

File: GameServerCore/Maps/IMapScriptHandler.cs [hunk 8]
```diff
@@ -124,6 +128,15 @@ public interface IMapScriptHandler : IUpdate
         /// </summary>
         /// <returns></returns>
         float GameTime();
-        void LoadBuildingProtection();
+        /// <summary>
+        /// Returns a list of all players present
+        /// </summary>
+        /// <returns></returns>
+        List<IChampion> GetPlayers();
```
Comment (lzardy) [inline GameServerCore/Maps/IMapScriptHandler.cs]: This function is unnecessary as scripts have access to ApiFunctionManager.

**Review by lzardy — CHANGES_REQUESTED** [review]
Few changes needed, looks good otherwise.

**Review by lzardy — APPROVED** [review]
Approved.

**Pull request merged**